### PR TITLE
r/datasync_location_object_storage

### LINKED ISSUE #22533: Support for self-managed object storage location
Hi. It would be great if this provider added support for the self-managed object storage as a location resource:
Original announcement: 

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### New or Affected Resource(s)

* aws_datasync_location_objectstorage

### Potential Terraform Configuration

```hcl

resource "aws_datasync_location_objectstorage" "example" {
  agent = aws_datasync_agent.example.name
  server = 1.1.1.1
  bucketname = "demo-datasync"
  Folder = /myfolder
  authentication = {
       access_key = var. key
       secret_key  = var.secretkey
   }
   } 
}
```
Please forgive any syntax errors. This is just a suggestion based on the GUI. This resource would then be consumed by the aws_datasync_task resource. The authentication block is seperate as we don't necessarily want to use our providers credentials for the sync itself. 
### References

* https://aws.amazon.com/blogs/aws/aws-datasync-adds-support-for-on-premises-object-storage/

### LINKED ISSUE #25633: Support Google Cloud Storage as a `aws_datasync_location`
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Hello, this is my first feature request, I hope I won't make mistakes.

AWS DataSync [now supports](https://aws.amazon.com/blogs/aws/new-for-aws-datasync-move-data-between-aws-and-google-cloud-storage-or-aws-and-microsoft-azure-files/) Google Cloud Storage (and Microsoft Azure Files) as locations. It would be nice to have it in Terraform

### New or Affected Resource(s)

* aws_datasync_location

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

None

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #22533
Closes #25633
Output from acceptance testing:

```
$ make testacc TESTS=TestAccDataSyncLocationObjectStorage_ PKG=datasync
--- PASS: TestAccDataSyncLocationObjectStorage_disappears (140.47s)
--- PASS: TestAccDataSyncLocationObjectStorage_basic (178.14s)
--- PASS: TestAccDataSyncLocationObjectStorage_tags (253.19s)
```

### CODE REVIEW

**Comment by DrFaust92** [comment]
I will not be looking more into this anytime soon, anyone can feel free to take over.

File: internal/service/datasync/sweep.go [hunk 1]
```diff
@@ -511,6 +516,63 @@ func sweepLocationHDFSs(region string) error {
 	return nil
 }
 
+func sweepLocationObjectStorages(region string) error {
```
Comment (DrFaust92) [inline internal/service/datasync/sweep.go:519]: TODO: refactor sweeper to new format + reuse one sweeper logic for all locations

**Comment by johnjelinek** [comment]
@DrFaust92: is this good to go? It looks good to me.

**Comment by johnjelinek** [comment]
@YakDriver: can we please get a review here?

**Comment by DrFaust92** [comment]
@johnjelinek, good to go

**Comment by johnjelinek** [comment]
@ewbankkit: can this one be reviewed?

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccDataSyncLocationObjectStorage_' PKG=datasync ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/datasync/... -v -count 1 -parallel 3  -run=TestAccDataSyncLocationObjectStorage_ -timeout 180m
=== RUN   TestAccDataSyncLocationObjectStorage_basic
=== PAUSE TestAccDataSyncLocationObjectStorage_basic
=== RUN   TestAccDataSyncLocationObjectStorage_disappears
=== PAUSE TestAccDataSyncLocationObjectStorage_disappears
=== RUN   TestAccDataSyncLocationObjectStorage_tags
=== PAUSE TestAccDataSyncLocationObjectStorage_tags
=== CONT  TestAccDataSyncLocationObjectStorage_basic
=== CONT  TestAccDataSyncLocationObjectStorage_tags
=== CONT  TestAccDataSyncLocationObjectStorage_disappears
--- PASS: TestAccDataSyncLocationObjectStorage_disappears (131.19s)
--- PASS: TestAccDataSyncLocationObjectStorage_basic (134.17s)
--- PASS: TestAccDataSyncLocationObjectStorage_tags (206.48s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/datasync	211.659s
```

**Comment by ewbankkit** [comment]
@DrFaust92 Thanks for the contribution :tada: :clap:.

**Pull request merged**